crates/board/src/crypto.rs

```diff
@@ -158,8 +177,79 @@ impl crate::Supported for sha2::Sha256 {}
 impl crate::Supported for sha2::Sha384 {}
 
 #[cfg(feature = "internal-software-crypto-hmac")]
-impl<D: Support<bool> + Default + BlockSizeUser + Update + FixedOutput + HashMarker> Support<bool>
-    for hmac::SimpleHmac<D>
+impl<D: Support<bool> + Default + BlockSizeUser + Update + FixedOutputReset + HashMarker>
+    Support<bool> for hmac::SimpleHmac<D>
 {
     const SUPPORT: bool = D::SUPPORT;
 }
+
+/// Extends APIs with infallible operations to support errors.
+///
+/// Infallible operations may return incorrect values. Those results must not be used until
+/// `last_error()` is called and returns `Ok(())`, signifying that the return value is correct.
+#[cfg(any(feature = "internal-api-crypto-hash", feature = "internal-api-crypto-hmac"))]
+pub trait LastError {
+    /// Returns the last error of infallible operations.
+    fn last_error(&self) -> Result<(), Error>;
+}
+
+/// Wrapper API around Hash that calls last_error.
+#[cfg(feature = "internal-api-crypto-hash")]
+pub struct HashApi<T: Hash>(pub T);
+
+/// Wrapper API around Hmac that calls last_error.
+#[cfg(feature = "internal-api-crypto-hmac")]
+pub struct HmacApi<T: Hmac>(pub T);
+
+#[cfg(feature = "internal-api-crypto-hash")]
+impl<T: Hash> OutputSizeUser for HashApi<T> {
+    type OutputSize = T::OutputSize;
+}
+
+#[cfg(feature = "internal-api-crypto-hmac")]
+impl<T: Hmac> OutputSizeUser for HmacApi<T> {
+    type OutputSize = T::OutputSize;
+}
+
+#[cfg(feature = "internal-api-crypto-hash")]
+impl<T: Hash> HashApi<T> {
+    /// Constructor of HmacApi.
+    pub fn new() -> Result<Self, Error> {
+        let hash = T::default();
+        hash.last_error().map(|()| HashApi(hash))
+    }
+
+    /// Call update from RustCrypto API and then last_error.
+    pub fn update(&mut self, data: &[u8]) -> Result<(), Error> {
+        self.0.update(data);
+        self.0.last_error()
+    }
+
+    /// Call finalize_into_reset from RustCrypto API and then last_error.
+    pub fn finalize_into_reset(&mut self, out: &mut Output<Self>) -> Result<(), Error> {
+        self.0.finalize_into_reset(out);
+        self.0.last_error()
+    }
+}
```

Comment:
to me: think of a way to factor out these common code for both structs

Reply:
I agree it looks like duplicated code a bit. But given we only have 2 instances (hash and hmac) and don't expect to get more, this duplication is actually fine (and the gain in readability by having the duplication is worth the duplication).